### PR TITLE
KEM: fix key retrieval when public/secret key is missing

### PR DESCRIPTION
**Pull request by dkostic**
### Issues:
Resolves V845882454.

### Description of changes: 
Fixes key retrieval when public/secret key is missing. Specifically,
when `EVP_PKEY_get_raw_private_key` is called on a key that has
only the public key an error will be returned. The same when calling
`EVP_PKEY_get_raw_public_key` with a key that has only the
private key.

Also, got rid of the `has_secret_key` field in the `KEM_KEY`
structure because it is not necessary.

### Call-outs:
Point out areas that need special attention or support during the review process. Discuss architecture or design changes.

### Testing:
How is this change tested (unit tests, fuzz tests, etc.)? Are there any testing steps to be verified by the reviewer?

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license and 
the ISC license.

### CODE REVIEW

File: crypto/evp_extra/p_kem.c [hunk 1]
```diff
@@ -145,7 +143,7 @@ static int pkey_kem_decapsulate(EVP_PKEY_CTX *ctx,
   }
 
   KEM_KEY *key = ctx->pkey->pkey.kem_key;
-  if (!key->has_secret_key) {
+  if (key->secret_key == NULL) {
```
Comment (andrewhop) [inline crypto/evp_extra/p_kem.c:146]: Should we add a similar safety check to encapsulate?

**Review by torben-hansen — COMMENTED** [review]
lgtm, except https://github.com/aws/aws-lc/pull/855/files#r1128788636 makes sense for encaps.

Do we check `shared_secret` anywhere for encaps?
I tried chasing it down the call-stack. But the Kyber reference implementation doesn't exactly make this an easy exercise.

File: crypto/evp_extra/p_kem.c (hunk 1 above)
Comment (dkostic) [reply crypto/evp_extra/p_kem.c:153]: added

Comment (dkostic) [reply crypto/evp_extra/p_kem.c:153]: btw, I'm working on a PR to add a bunch of test cases for failure modes of the API. Will add a test for this particular case as well.

**Review by andrewhop — APPROVED** [review]

**Review by torben-hansen — APPROVED** [review]

**Pull request merged**